File: algorithms/0003-longest-substring-without-repeating-characters/Solution.java
// Longest Substring Without Repeating Characters - sliding window + HashSet.
// Time O(n), Space O(min(n, alphabet)).

import java.util.HashSet;
import java.util.Set;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> seen = new HashSet<>();
        int left = 0;
        int best = 0;
        for (int right = 0; right < s.length(); right++) {
            char ch = s.charAt(right);
            while (seen.contains(ch)) {
                seen.remove(s.charAt(left));
                left++;
            }
            seen.add(ch);
            best = Math.max(best, right - left + 1);
        }
        return best;
    }
}

File: algorithms/0003-longest-substring-without-repeating-characters/solution.py
# Longest Substring Without Repeating Characters - sliding window + set.
# Time O(n), Space O(min(n, alphabet)).

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen: set[str] = set()
        left = 0
        best = 0
        for right, ch in enumerate(s):
            while ch in seen:
                seen.remove(s[left])
                left += 1
            seen.add(ch)
            best = max(best, right - left + 1)
        return best
